src/prof_recent.c

```diff
@@ -506,33 +575,70 @@ prof_recent_alloc_dump(tsd_t *tsd, void (*write_cb)(void *, const char *),
 	emitter_init(&emitter, emitter_output_json_compact,
 	    buf_writer_get_write_cb(&buf_writer),
 	    buf_writer_get_cbopaque(&buf_writer));
+
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_dump_mtx);
 	emitter_begin(&emitter);
 
-	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
-	prof_recent_alloc_assert_count(tsd);
+	ssize_t dump_count, dump_max;
+	prof_recent_t *dump_dummy = prof_recent_alloc_swap_and_clear_locked(
+	    tsd, &dump_count, &dump_max);
 
-	/*
-	 * Set prof_recent_alloc_max to 0 so that dumping won't block sampled
-	 * allocations: the allocations can complete but will not be recorded.
-	 */
-	ssize_t max = prof_recent_alloc_max_update(tsd, 0);
+	emitter_json_kv(&emitter, "recent_alloc_max", emitter_type_ssize,
+		&dump_max);
+	emitter_json_array_kv_begin(&emitter, "recent_alloc");
 
-	emitter_json_kv(&emitter, "recent_alloc_max", emitter_type_ssize, &max);
+	buf_writer_set_manual(&buf_writer);
+	prof_recent_t *dump_last = NULL;
+	prof_recent_t *node = NULL;
+	while (true) {
+		malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+		if (node == NULL) {
+			node = dump_dummy->next;
+		}
+		assert(node != NULL);
+		bool need_more_batch = false;
+		while (node != dump_dummy) {
+			buf_writer_commit(&buf_writer);
+			prof_recent_alloc_dump_node(&emitter, node);
+			if (buf_writer_overflow(&buf_writer)) {
+				buf_writer_revert(&buf_writer);
+				if (buf_writer_empty(&buf_writer)) {
+					buf_writer_set_auto(&buf_writer);
+					prof_recent_alloc_dump_node(
+					    &emitter, node);
+					buf_writer_set_manual(&buf_writer);
+				} else {
+					need_more_batch = true;
```

Comment:
If we do unlock-flush-relock right here, it should simplify the logic?

Reply:
Right, though such a change would look exactly like a revert of your #1763? :)

Reply:
Yeah however that one was to enable flat combining and reduce locking for the bin lock (which is heavily contended); here the recent lock is on a slow-slow path (i.e. dumping), and there won't be multiple threads going the same path (because of the dump lock). So no need to really pay the complexity.

Reply:
It's true that only one thread goes through the dump logic at a time because of the dump lock, but the `prof_recent_alloc_mtx` can be hot, since every thread needs it whenever there's a sampled allocation / deallocation.

Reply:
Discussed offline. The dump path won't be flat-combined by other threads, so we don't care about the lock pattern here.

Reply:
Revised. The logic looks much simpler now!